### DISCUSSION TITLE
How can I get a mutable state within a fairing?

### DISCUSSION THREAD

**Discussion by asnaseer-resilient** [rwf2/Rocket#2398] [Questions] [answered]
I am using Rocket v0.5.0-rc.1 and have added a managed state to rocket using `rocket.manage(my_state)`. I also have a fairing that is executed during liftoff and I would like to update something in `my_state` within this fairing - something like this:
```
rocket::build()
.mount(...)
.attach(AdHoc::on_ignite("On ignite", move |rocket| async move {
  ...
  // create my_state here
  ...
  rocket.manage(my_state)
}))
.attach(AdHoc::on_liftoff("On liftoff", |rocket| {
  match rocket.state::<MyState>() {
    None => warn!("MyState not available during liftoff?!"),
    Some(my_state) => my_state.some_mutable_method(some_data),
  }
}))
```
The `my_state` variable within this fairing an immutable reference to the state. However, I would like to call a mutable method on it and would therefore like a mutable reference to it.

Is there anyway of getting this?

**Comment by the10thWiz** [comment, marked as answer]
Since Fairings can run in parallel, you Rocket can't give you mutable access to the managed state. As such, you would need to use some kind of interior mutability, such as a Mutex or RwLock.
